src/mlpack/methods/ann/layer/isru.hpp

```diff
@@ -0,0 +1,183 @@
+   * f(x) by propagating x backwards through f. Using the results from the feed
+   * forward pass.
+   * 
+   * @param input The propagated input activation f(x).
+   * @param gy The backpropagated error.
+   * @param g The calculated gradient
+   */
+  template<typename DataType>
+  void Backward(const DataType&& input,
+                DataType&& gy,
+                DataType&& g);
+
+  //! Get the output parameter.
+  OutputDataType const& OutputParameter() const { return outputParameter; }
+  //! Modify the output parameter.
+  OutputDataType& OutputParameter() { return outputParameter; }
+
+  //! Get the delta.
+  OutputDataType const& Delta() const { return delta; }
+  //! Modify the delta.
+  OutputDataType& Delta() { return delta; }
+
+  //! Get the hyperparameter alpha.
+  double const& Alpha() const { return alpha; }
+  //! Modify the hyperparameter alpha.
+  double& Alpha() { return alpha; }
+
+  /**
+   * Serialize the layer.
+   */
+  template<typename Archive>
+  void serialize(Archive& ar, const unsigned int /* version */);
+
+ private:
+  /**
+   * Computes the value of ISRU activation function.
+   *
+   * @param x Input data.
+   * @return f(x) 
+   */
+  double Fn(const double x)
+  {
+    return x / (std::sqrt(1 + alpha*x*x));
+  }
+
+  /**
+   * Computes the value of ISRU activation function using a dense matrix
+   * as input.
+   * 
+   * @param x Input data.
+   * @param y The resulting output activation.
+   */
+  template<typename eT>
+  void Fn(const arma::Mat<eT>& x, arma::Mat<eT>& y)
+  {
+    y = x / (arma::sqrt(1 + alpha * arma::pow(x, 2)));
+  }
+
+  /**
+   * Computes the inverse of the ISRU function
+   * 
+   * @param y
+   * @return f^{-1}(y)
+   */
+  double Inv(const double y)
+  {
+    return y / std::sqrt(1 - alpha * y * y);
+  }
+
+  /**
+   * Computes the inverse of the ISRU function
+   * 
+   * @param y Input data.
+   * @param x The resulting inverse of the input data
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void Inv(const InputVecType& y,
+                  OutputVecType& x,
+                  const double alpha = 0.1)
+  {
+    x = y / arma::sqrt(1 - alpha * arma::pow(y, 2));
+  }
+
+  /**
+   * Computes the first derivate of the ISRU function
+   * 
+   * @param y Input activation
+   * @return f'(x) where f(x) = y
+   */
+  double Deriv(const double y)
+  {
+    if (y == 0)
+      return 1;
+    double x = Inv(y, alpha);
+    return std::pow(y / x, 3);
+  }
+
+  /**
+   * Computes the first derivative of the ISRU function.
+   *
+   * @param y Input activations.
+   * @param x The resulting derivatives. Should be the matrix used to calculate activation y 
+   * @param alpha parameter, default value = 0.1
+   */
+  template<typename InputVecType, typename OutputVecType>
+  void Deriv(const InputVecType& y,
+                    OutputVecType& x)
+  {
+    Inv(y, x, alpha);
+    x = arma::pow(y / x, 3);
+    x.replace(arma::datum::nan, 1);
```

Comment:
I still think we should avoid getting nan, incase any compiler has issue with it. You can write it as

`x = (x != 0) * arma::pow(y / x, 3) + (x == 0) * DBL_MAX;`

Reply:
Wow nice catch, I totally didn't consider the case when compiler may have problems with evaluating to nan.

Reply:
I think rather than this you can use for loop to make y(i) = Deriv(x(i))
and include this condition in Deriv condition in case you haven't already done so.